PI-941: Created <WaitForSettings> which only loads children if settin…

jwineman opened the pull request:
```
isEverythingLoaded && !isZoneOnCloudflare
```
Is missing but we don't have to code for it because its not possible for everything to be loaded if a zone is not provisioned on Cloudflare.

thellimist commented on src/js/components/WaitForSettings/WaitForSettings.js:
```diff
@@ -0,0 +1,52 @@
+  render() {
+    let activeZone = this.props.activeZone;
+    let analytics = (typeof this.props.analytics != 'undefined' ? this.props.analyics : true);
+    let settings = (typeof this.props.settings != 'undefined' ? this.props.settings : true);
```
should these be `'undefined'` as strings not types?

jwineman replied:
I'm confused by your comment because putting `undefined` in single quotes makes it a string and [`typeof` returns a string](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof).

The selectors we use in HomePage, MoreSettingsPage, and AnalyticsPage return `false` if the various settings aren't found.  If the prop is `undefined` that means it wasn't passed in and we don't need to check it so we set it to `true`.

thellimist replied:
I meant the opposite. 

should these *not* be `'undefined'` as strings *but defined as*  not `undefined`.

Currently there are single quotes which needs to be removed I think

jwineman replied:
```
typeof variableThatDoesntExist
"undefined"
```
Notice it returns the string `"undefined"` not the type `undefined`.

Your comment below about switching is essentially how this is implemented.  Non-empty objects will evaluate to `true` when we evaluate:
```
 let isEverythingLoaded = settings && pluginSettings && analytics;
```

for this:
```
let analytics = (typeof this.props.analytics != 'undefined' ? this.props.analyics : true);
```
If its not `undefined` it exists, and we will use the object which will evaluate to `true` or it will be the value `false` returned from the selector if it can't find the object.
If it is `undefined` that means it wasn't passed in as a prop and we want to set it to `true` since it doesn't need to be checked.

jwineman replied:
The reason I didn't do only booleans is because we would have to write 3 more selectors that just check the existence of stuff rather than returning it.

jwineman replied:
If you feel strongly about it I can rework it but I figured it was fine since `{...}` evaluates to `true`

thellimist replied:
This is not readable. The viewer should look inside the view to understand whats going on.
`<WaitForSettings activeZone={activeZone} settings={getAllZoneSettingsForZoneId(activeZone.id, zoneSettings)} pluginSettings={getPluginSettingsForZoneId(activeZone.id, pluginSettings)}>`

That is why I suggested a different blueprint such as
`<WaitForSettings waitZone=true waitSettings=true waitPluginSettings=true waitAnalytics=false/>`

I'm not a fan of implicit code. We can make `WaitForSettings` really dumb and calculate all of the fields in every page explicitly 

```
let shouldWaitAnalytics = (typeof this.props.analytics == 'undefined' ? true : false);
let shouldWaitSettings = (typeof this.props.settings == 'undefined' ? true : false);
let shouldWaitPluginSettings = (typeof this.props.pluginSettings == 'undefined' ? true : false);
```

Afterwards the blueprint will like

``<WaitForSettings waitZone={shouldWaitZone} waitSettings={shouldWaitSettings} waitPluginSettings={shouldWaitPluginSettings} waitAnalytics=false}/>`` (for non anaytic pages)

thellimist commented on src/js/components/WaitForSettings/WaitForSettings.js:
```diff
@@ -0,0 +1,52 @@
+  ]),
+  analytics: React.PropTypes.oneOfType([
+    React.PropTypes.object,
+    React.PropTypes.oneOf([false])
```
I think this class could be better if the definition is like

`<WaitForSettings waitZone=true waitSettings=true waitPluginSettings=true waitAnalytics=false/>`

This way we can abstract everything and only pass booleans to describe what to wait. It'll need to implicitly grab the `activeZone`.

thellimist commented:
Undefined syntax:

http://prntscr.com/e19kel

jwineman commented:
You aren't using `typeof`:
<img width="260" alt="screen shot 2017-01-27 at 4 20 15 pm" src="https://cloud.githubusercontent.com/assets/6947158/22392315/8db41bf6-e4ac-11e6-8e3c-ca57f1ad99db.png">

You're right though, the code is not easy to read.  I'll rework it to be explicitly boolean values.

thellimist commented:
Ohh sorry missed that. I never assumed we'd check `undefined` value with `type of`. We can remove `type of` since we don't need it.

jwineman commented:
In the current implementation I need it to distinguish between `false` and `undefined`

thellimist commented:
Why?
<img width="147" alt="screen shot 2017-01-28 at 00 55 12" src="https://cloud.githubusercontent.com/assets/6690196/22392726/74f5a894-e4f4-11e6-8892-4a3f8b5342c7.png">

jwineman commented:
Looks like the diff on `AnalyticsPage` is messed up.  If you have:
```
<WaitForSettings>
{analytics.totals.requests.all} 
<WaitForSettings>
```
React will try to evaluate `analytics.totals.requests.all` even if `<WaitForSettings>` decides not to display it. If the request hasn't completed it yet it will be `undefined` and the whole thing will error out.  The fix is:
```
<WaitForSettings>
{analytics.totals ? analytics.totals.requests.all : 0 } 
<WaitForSettings>
```
So the changes are:

1. Moved the time series chart into `<TimeSeriesChart>`
2. Added ternary checks for all the `analytics` data `AnalyticsPage` tries to display
3. Created functions for the logic around column data, and top threat etc.

thellimist commented on src/js/components/TimeSeriesChart/TimeSeriesChart.js:
```diff
@@ -0,0 +1,63 @@
+    //The first element in the remaining rows column has to be the label
+    for(let i = 0; i < xAxisValues.length; i++) {
+      if(xAxisValues[i].values) {
+        columns.push([xAxisValues[i].label].concat(xAxisValues[i].values));
```
`columns.push([xAxisValues[i].label].concat(xAxisValues[i].values));`

It's hard to understand the paranthesis. Can we write it more explicit as in `xAxisValues[i].label` in variable etc. to make it easier to read

jwineman replied:
16e79b0d6289433b421f401f45036195c6452ff9

thellimist commented on src/js/containers/HomePage/HomePage.js:
```diff
@@ -1,47 +1,28 @@
-        );
-    }
+    return (
+      <WaitForSettings settings pluginSettings>
```
this looks really clean :thumbsup:

We need this in MoreSettingsPage, Analytic Page and DNSManagement Page.

jwineman replied:
The MoreSettingsPage and AnalyticsPage have `<WaitForSettings>` as part of this PR already, I think its in the previous commit though.

I'll check the DNS management Page but I think it checks already, if not I'll add it.

thellimist replied:
LGTM
